server/interceptor/interceptor.go

```diff
@@ -0,0 +1,170 @@
+// Copyright 2017 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package interceptor defines gRPC interceptors for Trillian.
+package interceptor
+
+import (
+	"strings"
+
+	"github.com/google/trillian"
+	"github.com/google/trillian/errors"
+	"github.com/google/trillian/storage"
+	"github.com/google/trillian/trees"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+)
+
+// TreeInterceptor ensures that all requests pertaining a specific tree are of correct type and
+// respect the current tree state (frozen, deleted, etc).
+type TreeInterceptor struct {
+	Admin storage.AdminStorage
+}
+
+// UnaryInterceptor executes the TreeInterceptor logic for unary RPCs.
+func (i *TreeInterceptor) UnaryInterceptor(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+	rpcInfo, err := getRPCInfo(req, info.FullMethod)
+	if err != nil {
+		return nil, err
+	}
+
+	if rpcInfo.doesNotHaveTree {
+		return handler(ctx, req)
+	}
+
+	tree, err := trees.GetTree(ctx, i.Admin, rpcInfo.treeID, rpcInfo.opts)
+	if err != nil {
+		return nil, err
```

Comment:
Hmm, good call. Since this is a server/gRPC layer, I've replaced usages of `trillian/errors` with (the newly-added) `grpc/status`.

I'll see about interceptor-level wrapping in a separate PR, as we could potentially solve *all* of our wrapping problems that way.